### PR TITLE
support image_embs input

### PR DESCRIPTION
**Pull request by irexyc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/turbomind/turbomind.py [hunk 1]
```diff
@@ -459,6 +459,8 @@ async def async_stream_infer(self, *args, **kwargs):
     def stream_infer(self,
                      session_id,
                      input_ids,
+                     image_embs=None,
```
Comment (lvhan028) [inline lmdeploy/turbomind/turbomind.py:462]: please update docstring

File: lmdeploy/turbomind/turbomind.py [hunk 2]
```diff
@@ -544,6 +546,29 @@ def _broadcast_np(data, dtype, shape=(batch_size, )):
             CORRID=np.array(session_id, dtype=np.uint64),
             STOP=_broadcast_np((1 if stop else 0), np.int32))
 
+        if image_embs is not None:
+            assert len(image_offsets) == len(image_embs)
+            # image_embs     Union[List[np.array], List[List[np.array]]]
```
Comment (lvhan028) [inline lmdeploy/turbomind/turbomind.py:551]: What's the case for `List[List[np.array]]` and `List[List[int]]`

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 3]
```diff
@@ -258,6 +258,30 @@ void LlamaBatch<T>::ProcessInferRequests(const Requests& requests)
             output_ids = Copy(input_ids, input_length, output_ids);
         }
 
+        // copy image embeddings
+        if (model_->image_dim_ > 0 && r->inputs[rank_].isExist("image_embs")) {
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaBatch.cc:262]: When will `model_->image_dim_ >0`?

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 4]
```diff
@@ -258,6 +258,30 @@ void LlamaBatch<T>::ProcessInferRequests(const Requests& requests)
...
+                TM_LOG_WARNING("[ImageFeature] Invalid image feature, id = %ld, info = %s",
+                               (long)seq.id,
+                               image_embs_tensor.toString().c_str());
+                continue;
+            }
+
+            T*         image_embs      = r->inputs[rank_].getPtr<T>("image_embs");
+            const int* h_image_offsets = r->inputs[rank_].getPtr<int>("image_offsets");
+            const int  count           = model_->image_dim_ * model_->hidden_units_;
+            for (size_t i = 0; i < n_offsets && h_image_offsets[i] > 0; i++) {
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaBatch.cc:276]: Can `h_image_offfsets[i]` be `0`?

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 5]
```diff
@@ -258,6 +258,30 @@ void LlamaBatch<T>::ProcessInferRequests(const Requests& requests)
...
+                seq.image_offsets.push_back(seq.tokens.size() + h_image_offsets[i]);
+                auto& emb = seq.image_embs.emplace_back();
+                emb.resize(count * sizeof(T));
+                std::memcpy(emb.data(), image_embs, count * sizeof(T));
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaBatch.cc:280]: If assume `image_embs` is a CPU tensor, I think it's better to check it in `turbomind.py`

Comment (lvhan028) [reply src/turbomind/models/llama/LlamaBatch.cc:280]: @lzhangzz does turbomind request all user's inputs being on CPU?

File: lmdeploy/turbomind/deploy/target_model/base.py [hunk 6]
```diff
@@ -58,6 +58,7 @@ class TurbomindModelConfig:
     max_position_embeddings: int = 0
     rope_scaling_factor: float = 0.0
     use_logn_attn: int = 0
+    image_dim: int = 0
```
[the remark this answers is not in the document]
Comment (irexyc) [reply lmdeploy/turbomind/deploy/target_model/base.py:61]: 不是，没有固定的格式。
qwen-vl是https://huggingface.co/Qwen/Qwen-VL/blob/main/tokenization_qwen.py#L50这里。
internlm-xcomposer 有这个字段 https://huggingface.co/internlm/internlm-xcomposer-vl-7b/blob/main/config.json#L48，但是image feature首尾得加上特殊的embedding，所以得加2

File: lmdeploy/turbomind/turbomind.py (hunk 2 above)
Comment (irexyc) [reply lmdeploy/turbomind/turbomind.py:551]: 跟input_ids一致，input_ids可以是List[int], List[List[int]]。之前这里可以传batch的输入，现在好像不行了。

File: src/turbomind/models/llama/LlamaBatch.cc (hunk 3 above)
Comment (irexyc) [reply src/turbomind/models/llama/LlamaBatch.cc:262]: 目前是手动设置的。

File: src/turbomind/models/llama/LlamaBatch.cc (hunk 5 above)
Comment (lzhangzz) [reply src/turbomind/models/llama/LlamaBatch.cc:280]: > @lzhangzz does turbomind request all user's inputs being on CPU?

I prefer to keep them in host as long as it doesn't impose unavoidable performance penalty.

File: lmdeploy/turbomind/turbomind.py [hunk 7]
```diff
@@ -544,6 +548,29 @@ def _broadcast_np(data, dtype, shape=(batch_size, )):
...
+            image_embs = [[
+                torch.from_numpy(x).squeeze().unsqueeze(0) for x in y
+            ] for y in image_embs]
+            image_embs = [torch.cat(x) for x in image_embs]
+            image_embs = pad_sequence(image_embs, batch_first=True)
+            image_offsets = [torch.IntTensor(x) for x in image_offsets]
+            image_offsets = pad_sequence(image_offsets,
+                                         batch_first=True,
+                                         padding_value=-1)
+            if self.tm_model.config.weight_type == 'fp32':
```
[the remark this answers is not in the document]
Comment (irexyc) [reply lmdeploy/turbomind/turbomind.py:562]: int4，lookup table也是float的吧

File: src/turbomind/models/llama/LlamaBatch.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (lvhan028) [reply src/turbomind/models/llama/LlamaBatch.cc]: 是不是用 input_embeddings 更好些啊

[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/models/llama/LlamaBatch.cc]: input_embeddings 感觉有点像transformers里面的inputs_embeds，跟input_ids是二选一的关系

**Review by lvhan028 — APPROVED** [review]

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 8]
```diff
@@ -258,6 +275,53 @@ void LlamaBatch<T>::ProcessInferRequests(const Requests& requests)
             output_ids = Copy(input_ids, input_length, output_ids);
         }
 
+        // copy input embeddings
+        if (r->inputs[rank_].isExist("embedding_counts")) {
+            int        emb_count    = r->inputs[rank_].getVal<int>("embedding_counts");
+            const auto emb_tensor   = r->inputs[rank_].at("input_embeddings");
+            const auto begin_tensor = r->inputs[rank_].at("embedding_begins");
+            const auto end_tensor   = r->inputs[rank_].at("embedding_ends");
```
Comment (lzhangzz) [inline src/turbomind/models/llama/LlamaBatch.cc:283]: It is overly verbose to use 3 parameters for just specifying the range of input embeddings.

Use a single `input_embedding_ranges` with shape `[batch_size, max_input_embedding_count, 2]`, store the `begin`/`end` pairs in the inner most dimension and pad to `max_input_embedding_count` with `(-1, -1)`

**Review by lzhangzz — APPROVED** [review]

**Pull request merged**